Implement part unlocks

0HyperCube opened the pull request:
**Brief Description of What This PR Does**
- [x] Unlock system
- [x] Some basic unlock conditions
- [x] Show progress towards unlocking
- [x] Unlock all in freebuild mode
- [x] Show undiscovered icon
- [ ] Use translation system

**Related Issues**
Closes #3935

**Progress Checklist**
Note: before starting this checklist the PR should be marked as non-draft.

- [ ] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

hhyyrylainen commented:
I feel like this might be a bit early(?) to try to implement this before we even have: https://github.com/Revolutionary-Games/Thrive/issues/3935 ...

hhyyrylainen commented:
I saw that you started to include the organelle unlock system in this PR. The reason why I made that comment is that you have basically jumped into the deep end of the pool (as shown by the various half-discussed points you referred to about the unlocks). Much easier on-ramp to this would have been to first implement the organelle unlock system, where the first unlock condition to implement would be to unlock the thylakoid only once the player reaches a patch that has light. And also that feature asks for tooltips and hints for the player to see easily how to unlock more stuff. So basically to get this PR done, you need to do *2* massive systems at the same time...

> Another proposed method of unlocking organelles was "spend points in the editor to randomly unlock a new Protein Organelle" - which I'm happy to add here if you wish? If so, do you have any preferences for where that should be in the UI?

This PR is going to be scrutinized a lot by the gameplay design team (and hopefully other team members as well) so I can't say for sure. We'll need playtesting results to determine how the initial implementation needs to be tweaked to make it work.

> This PR also a HUD message when a new organelle is unlocked after you engulf something. Although perhaps there should also be a hint in the editor for this? If so, where should this be?

I'd say it is not enough as the player has forgotten by the time they get to the editor. We should have a some kind of notice message or unlocking animation that starts when the player gets to the microbe editor.

Deus-Codes commented:
Impressive that you are tackling this, but I'd also repeat hh's comments to make sure to be aware of current designs to avoid having to dramatically change finished work. To clarify, this is the current conceptual understanding of how endosymbiosis and unlocks will work: 

https://forum.revolutionarygamesstudio.com/t/upgrades-unlocks-endosymbiosis-master-thread/915/1

It's a long thread, so please contact one of us if you have any questions.

hhyyrylainen commented:
I'll add that that thread ended up being mostly about upgrades. For unlocks I think the key concepts are:
- a general system for keeping track of unlock conditions for organelles (this should be easily extendable with new conditions)
- GUI work to hide locked organelles and show progress towards the unlock conditions
- Some organelles are always unlocked
- We will probably have multiple alternative ways to unlock an organelle (for example endosymbiosis and some optional condition)

There's a separate thread discussing *just* unlocks here, which I've linked in the organelle unlock system issue:

https://forum.revolutionarygamesstudio.com/t/organelle-unlocks/696

Deus-Codes commented:
@0HyperCube if you wish, you can tag me or any member of the game design team on the community discord/forum to discuss implementation since you aren't in the dev team currently.

0HyperCube commented:
I have added some basic support for the unlocking behaviour that you specified:

> a general system for keeping track of unlock conditions for organelles (this should be easily extendable with new conditions)

I've added this along with the proposed unlock conditions from the [first post of one of the threads](https://forum.revolutionarygamesstudio.com/t/upgrades-unlocks-endosymbiosis-master-thread/915#unlocks-1). Values and conditions can be easily changed from the `organelles.json` fle.

> GUI work to hide locked organelles and show progress towards the unlock conditions

I'm not quite sure how to show the progress if the organelles are hidden (as I currently store the progress in a tooltip) - do you have any suggestions for this?

> Some organelles are always unlocked

Organelles where the `unlockConditions` is not specified in the `organelles.json` file are unlocked by default.

> We will probably have multiple alternative ways to unlock an organelle (for example endosymbiosis and some optional condition)

Endosymbiosis continues to work - although I can remove it from this PR if that would make it easier to review? Endosymbiosis alone is relatively few lines though.

![Demo of locked organelles](https://github.com/Revolutionary-Games/Thrive/assets/78500760/f1534200-2c85-4280-8b06-9e769798dc88)

Any feedback would be most appreciated!

Deus-Codes commented:
I would like to play test this more, but I really like the way information is displayed in the tooltips and the unlock system as a whole. As expected, it adds a nice layer of depth and replayability to the game which is something we really needed. 

I will say that I think adding endosymbiosis with this would be a rather big change, and considering that how you implemented endosymbiosis doesn't match up with our current concepts, it might be best just to limit this pull request to the unlock system (which is incredible enough in itself). Don't do this yet as this is my personal opinion and perhaps other team members have a different take, but I'm leaning towards just having the basic upgrade system implemented in this pull personally. 

The GUI tooltip is really good in being dynamic, letting the player know what needs to be done and their current status. It might be beneficial to add some emphasis within the text so that players can more easily find it. 

Regarding the editor buttons that Hh mentioned previously, the ideal situation is to have parts hidden and have a question mark button indicating to the player that there are undiscovered parts still available to minimize clutter. This button would be present at the end of a part section of the editor until all parts are discovered. Check out the last visual in this post (https://forum.revolutionarygamesstudio.com/t/organelle-unlocks/696/34). 

One other thing I noticed, unlock conditions apply to free build where we should probably give players access to all parts instantly to let them experiment and have fun. 

Incredible work overall, it really made me happy to open the game and see unlock conditions. I'll likely have more input later so you're not off the proverbial hook just yet, but thank you for volunteering your time.

hhyyrylainen commented:
> I'm not quite sure how to show the progress if the organelles are hidden (as I currently store the progress in a tooltip) - do you have any suggestions for this?

The GDD has a GUI mockup: https://wiki.revolutionarygamesstudio.com/wiki/Microbe_Stage_GDD

Basically replacing any hidden organelles in one category with just a single "?" icon with a tooltip showing the unlock progress towards the next organelle, is the design that's got the most support behind it. [Direct link to the mockup](https://wiki.revolutionarygamesstudio.com/wiki/File:Locked_Organelles.jpg).

> I will say that I think adding endosymbiosis with this would be a rather big change, and considering that how you implemented endosymbiosis doesn't match up with our current concepts, it might be best just to limit this pull request to the unlock system (which is incredible enough in itself). Don't do this yet as this is my personal opinion and perhaps other team members have a different take, but I'm leaning towards just having the basic upgrade system implemented in this pull personally.

I did say it before, but in my opinion doing the organelle unlock system and endosymbiosis is pretty crazy to try to do in a single PR. So I'd also say that it'd make things much easier and faster if this first PR just adds the unlock system and then a second PR is made for the endosymbiosis feature.

> One other thing I noticed, unlock conditions apply to free build where we should probably give players access to all parts instantly to let them experiment and have fun.

Yes, freebuild should be excluded from the unlock system (as it wouldn't be "free" build anymore).

0HyperCube commented:
Thanks for the feedback @Deus-Codes and @hhyyrylainen. I've fixed the freebuild, removed endosymbiosis and modified the organelle unlocks to be like the provided design as suggested.
![Undiscovered organelles tooltip](https://github.com/Revolutionary-Games/Thrive/assets/78500760/cad34ae2-1755-42dc-8c53-30f178d0b667)

hhyyrylainen commented:
I think that looks pretty good, and what I assume from the single image, it probably is close to what the initial unlock system version should be. I'll review the code later once the functionality is about complete.

0HyperCube commented:
Other than the translations, is there any more functionality needed @hhyyrylainen?

hhyyrylainen commented:
In case you didn't see it, we have documentation on the translations: https://github.com/Revolutionary-Games/Thrive/blob/master/doc/working_with_translations.md

I have not playtested to check. @Deus-Codes do you have time to playtest this again?

buckly90 commented:
Wonderful work, wonderful! It's precisely as envisioned minus the need for some graphics tweaks but that's a battle for our artists.
I believe this will go a long way in helping new players with the learning curve of the game.

The only real issue I noticed was that currently the nitrogenase parts are utterly impossible to unlock because ammonia levels never drop that low.  Instead, the amount needed to unlock should be 13 or lower.

Another, albeit minor, issue is how the unlock conditions are presented. Currently it seems as if it is a list of tasks the player must complete to unlock the part. But in truth only one task need be completed.  adding "Or" between each task might mitigate this risk for confusion.

hhyyrylainen commented on simulation_parameters/microbe_stage/organelles.json:
```diff
@@ -22,6 +22,14 @@
     "iconPath": "res://assets/textures/gui/bevel/parts/PilusIcon.png",
+    "unlockConditions": [
+      {
+        "EngulfedMicrobesAbove": 5
```
You've correctly seen the new style for our JSON files here, but this file is old so it uses camelCase. To maintain file uniformity and make it less surprising to read this file, could you convert the parts you added to camelCase (`engulfedMicrobesAbove`)?

hhyyrylainen commented on src/microbe_stage/editor/CellEditorComponent.GUI.cs:
```diff
@@ -2,7 +2,10 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.Linq;
+using AngleSharp.Common;
```
What's used from AngleSharp here? AngleSharp is a HTML parsing library so I'd really prefer to keep it entirely outside gameplay code.

hhyyrylainen commented on src/microbe_stage/editor/MicrobePartSelection.tscn:
```diff
@@ -96,9 +96,24 @@ mouse_filter = 2
+[node name="Label" type="Label" parent="VBoxContainer/Button/RecentlyUnlocked"]
+margin_right = 30.0
+margin_bottom = 14.0
+text = "NEW"
```
I think the translation key should be more specific here like "UNLOCKED_NEW_ITEM". Also the uppercase specification here is something that isn't used elsewhere in the game, rather all the text uppercasing is handled with translations.

hhyyrylainen commented on src/microbe_stage/organelle_unlocks/IUnlockConstraint.cs:
```diff
@@ -0,0 +1,211 @@
+    /// <summary>
+    ///   The number of microbes engulfed by the player is above the specified amount.
+    /// </summary>
+    public struct EngulfedMicrobesAbove : IUnlockConstraint
```
Structs should *only* be used when it really matters for performance. Our default style is that everything is a class unless can shown otherwise that it is needed.

hhyyrylainen commented on src/microbe_stage/organelle_unlocks/OrganelleUnlockConstraints.cs:
```diff
@@ -0,0 +1,108 @@
+    private IEnumerable<IUnlockConstraint> UnlockConstraints()
+    {
+        if (EngulfedMicrobesAbove is not null)
+            yield return new EngulfedMicrobesAbove { Microbes = EngulfedMicrobesAbove.Value };
```
Unless I misunderstood C# boxing rules, the use of structs is negated here with basically mandatory boxing to transfer the data through an enumerator out of this method. Also this creating of temporary objects to check the conditions isn't exactly my preferred method I'd design this. Instead even when loading the data from JSON the actual class types that would do the checking can be loaded directly. For example see how organelle component factories are specified for the OrganelleDefinition directly in the JSON.

hhyyrylainen commented on src/general/GameWorld.cs:
```diff
@@ -18,6 +18,15 @@
+    public UnlockProgress UnlockProgress = new();
+
+    [JsonProperty]
+    public int TotalPlayerDeaths;
```
When thinking about potential designs for the unlock system, I thought that the unlock progress tracking would be handled by the unlocks system itself because otherwise we end up with a ton of tracking variables in the world or/and the microbe class. So related to my previous comment I think the way to construct the unlock progress would be to create the unlock condition objects once for a started game and then store the state about the unlock conditions in those class instances. Feel free to suggest something else, but do you think a design like that could work?

0HyperCube replied:
I presumably need to store the total player deaths somewhere within the save file though?

hhyyrylainen replied:
Well with my suggestion you'd save the entire tree of unlock condition objects, and each of those would keep track of the statistics they care about. To simplify the design and make it extensible gameplay code should have helper methods made in the main unlock conditions class which then could forward game events to the relevant unlock condition objects (that way gameplay code is less tightly coupled to specific unlock condition implementations).

0HyperCube replied:
I'm not sure I have time or C# ability to rewrite this PR using that system, so I'll just close it.

It has been interesting playing around with this game and I'd like to thank you for your time and feedback.

hhyyrylainen replied:
From my perspective this was almost complete, so it is a bit unfortunate to see that this won't make it all the way to the finish line. But I guess that's how it is. If you want to come back in a couple of months to finish this, you are entirely welcome to do so. Or maybe someone else will be able to complete this feature. 
Anyway I want to also say thanks for at least that one PR you did complete and helped develop Thrive.
